File: app/src/main/java/com/example/myapplication/core/nutrition/NutritionTargetCalculator.kt
package com.example.myapplication.core.nutrition

import com.example.myapplication.core.profile.GoalPace
import com.example.myapplication.core.profile.MetabolicSex
import com.example.myapplication.core.profile.PersonalProfile
import kotlin.math.abs
import kotlin.math.min
import kotlin.math.roundToInt

class NutritionTargetCalculator {
    fun calculate(
        profile: PersonalProfile,
        ageYears: Int,
        timeline: TargetTimeline? = null,
    ): CalculationResult {
        validate(profile, ageYears)?.let { reason ->
            return CalculationResult.NeedsProfessionalReview(reason)
        }
        validateTimeline(profile, timeline)?.let { reason ->
            return CalculationResult.NeedsProfessionalReview(reason)
        }

        val rawBasalCalories =
            10 * profile.currentWeightKg +
                6.25 * profile.heightCm -
                5 * ageYears +
                if (profile.metabolicSex == MetabolicSex.MALE) MALE_CONSTANT else FEMALE_CONSTANT
        val rawMaintenanceCalories = rawBasalCalories * profile.activityLevel.multiplier
        val rawTargetCalories = when {
            profile.targetWeightKg < profile.currentWeightKg -> {
                val rate = when (profile.goalPace) {
                    GoalPace.MILD -> 0.10
                    GoalPace.STANDARD -> 0.15
                    GoalPace.AGGRESSIVE -> 0.20
                }
                rawMaintenanceCalories * (1.0 - rate)
            }
            profile.targetWeightKg > profile.currentWeightKg -> {
                val rate = when (profile.goalPace) {
                    GoalPace.MILD -> 0.05
                    GoalPace.STANDARD -> 0.10
                    GoalPace.AGGRESSIVE -> 0.15
                }
                rawMaintenanceCalories * (1.0 + rate)
            }
            else -> rawMaintenanceCalories
        }

        val rawProteinGrams = profile.currentWeightKg * PROTEIN_GRAMS_PER_KG
        val rawFatGrams = profile.currentWeightKg * FAT_GRAMS_PER_KG
        val rawCarbsGrams =
            (rawTargetCalories -
                rawProteinGrams * CALORIES_PER_PROTEIN_GRAM -
                rawFatGrams * CALORIES_PER_FAT_GRAM) / CALORIES_PER_CARB_GRAM

        if (
            rawBasalCalories <= 0 ||
            rawMaintenanceCalories <= 0 ||
            rawTargetCalories <= 0 ||
            rawCarbsGrams < 0
        ) {
            return CalculationResult.NeedsProfessionalReview(
                "Không thể tạo mục tiêu dinh dưỡng an toàn từ dữ liệu hiện tại.",
            )
        }

        return CalculationResult.Target(
            NutritionTarget(
                basalCalories = rawBasalCalories.roundToInt(),
                maintenanceCalories = rawMaintenanceCalories.roundToInt(),
                calories = rawTargetCalories.roundToInt(),
                proteinGrams = rawProteinGrams.roundToInt(),
                carbsGrams = rawCarbsGrams.roundToInt(),
                fatGrams = rawFatGrams.roundToInt(),
                audit = NutritionTargetAudit(
                    rawBasalCalories = rawBasalCalories,
                    rawMaintenanceCalories = rawMaintenanceCalories,
                    rawTargetCalories = rawTargetCalories,
                    rawProteinGrams = rawProteinGrams,
                    rawCarbsGrams = rawCarbsGrams,
                    rawFatGrams = rawFatGrams,
                ),
            ),
        )
    }

    fun capAutomaticCalorieDelta(currentCalories: Int, requestedDelta: Int): Int {
        if (currentCalories <= 0) return 0

        val fivePercentCap = (currentCalories * AUTOMATIC_CHANGE_RATE).roundToInt()
        val cap = min(fivePercentCap, MAXIMUM_AUTOMATIC_CALORIES)
        return requestedDelta.coerceIn(-cap, cap)
    }

    private fun validate(profile: PersonalProfile, ageYears: Int): String? = when {
        ageYears !in MINIMUM_AGE..MAXIMUM_AGE -> "Độ tuổi cần nằm trong khoảng 18 đến 100."
        !profile.heightCm.isFinite() || profile.heightCm !in MINIMUM_HEIGHT_CM..MAXIMUM_HEIGHT_CM ->
            "Chiều cao nằm ngoài phạm vi hỗ trợ."
        !profile.currentWeightKg.isFinite() || profile.currentWeightKg !in MINIMUM_WEIGHT_KG..MAXIMUM_WEIGHT_KG ->
            "Cân nặng hiện tại nằm ngoài phạm vi hỗ trợ."
        !profile.targetWeightKg.isFinite() || profile.targetWeightKg !in MINIMUM_WEIGHT_KG..MAXIMUM_WEIGHT_KG ->
            "Cân nặng mục tiêu nằm ngoài phạm vi hỗ trợ."
        !profile.personalizationConsent -> "Cần đồng ý sử dụng dữ liệu hồ sơ trước khi tính mục tiêu."
        else -> null
    }

    private fun validateTimeline(profile: PersonalProfile, timeline: TargetTimeline?): String? {
        if (timeline == null) return null

        val durationDays = timeline.targetDateEpochDay - timeline.todayEpochDay
        if (durationDays <= 0) return "Ngày mục tiêu cần nằm sau ngày hiện tại."

        val durationWeeks = durationDays / DAYS_PER_WEEK
        val kilogramsPerWeek = abs(profile.targetWeightKg - profile.currentWeightKg) / durationWeeks
        return if (kilogramsPerWeek > MAXIMUM_KILOGRAMS_PER_WEEK) {
            "Tốc độ thay đổi cân nặng vượt 0,9 kg mỗi tuần; hãy chọn thời hạn dài hơn hoặc trao đổi với chuyên gia."
        } else {
            null
        }
    }

    private companion object {
        const val MALE_CONSTANT = 5.0
        const val FEMALE_CONSTANT = -161.0
        const val PROTEIN_GRAMS_PER_KG = 2.0
        const val FAT_GRAMS_PER_KG = 0.8
        const val CALORIES_PER_PROTEIN_GRAM = 4.0
        const val CALORIES_PER_CARB_GRAM = 4.0
        const val CALORIES_PER_FAT_GRAM = 9.0
        const val AUTOMATIC_CHANGE_RATE = 0.05
        const val MAXIMUM_AUTOMATIC_CALORIES = 150
        const val MAXIMUM_KILOGRAMS_PER_WEEK = 0.9
        const val DAYS_PER_WEEK = 7.0
        const val MINIMUM_AGE = 18
        const val MAXIMUM_AGE = 100
        const val MINIMUM_HEIGHT_CM = 100.0
        const val MAXIMUM_HEIGHT_CM = 250.0
        const val MINIMUM_WEIGHT_KG = 30.0
        const val MAXIMUM_WEIGHT_KG = 350.0
    }
}

File: flutter/lib/core/nutrition/nutrition_target_calculator.dart
import 'dart:math';
import '../model/nutrition_models.dart';
import '../model/profile_models.dart';

class NutritionTargetCalculator {
  static const double _maleConstant = 5.0;
  static const double _femaleConstant = -161.0;
  static const double _proteinGramsPerKg = 2.0;
  static const double _fatGramsPerKg = 0.8;
  static const double _caloriesPerProteinGram = 4.0;
  static const double _caloriesPerCarbGram = 4.0;
  static const double _caloriesPerFatGram = 9.0;
  static const double _automaticChangeRate = 0.05;
  static const int _maximumAutomaticCalories = 150;
  static const double _maximumKilogramsPerWeek = 0.9;
  static const double _daysPerWeek = 7.0;
  static const int _minimumAge = 18;
  static const int _maximumAge = 100;
  static const double _minimumHeightCm = 100.0;
  static const double _maximumHeightCm = 250.0;
  static const double _minimumWeightKg = 30.0;
  static const double _maximumWeightKg = 350.0;

  CalculationResult calculate({
    required PersonalProfile profile,
    required int ageYears,
    TargetTimeline? timeline,
  }) {
    final profileIssues = _validate(profile, ageYears);
    if (profileIssues != null) {
      return CalculationResult.needsProfessionalReview(profileIssues);
    }
    final timelineIssues = _validateTimeline(profile, timeline);
    if (timelineIssues != null) {
      return CalculationResult.needsProfessionalReview(timelineIssues);
    }

    final double rawBasalCalories = 10 * profile.currentWeightKg +
        6.25 * profile.heightCm -
        5 * ageYears +
        (profile.metabolicSex == MetabolicSex.male
            ? _maleConstant
            : _femaleConstant);

    final rawMaintenanceCalories =
        rawBasalCalories * profile.activityLevel.multiplier;

    final double rawTargetCalories;
    if (profile.targetWeightKg < profile.currentWeightKg) {
      final double rate;
      switch (profile.goalPace) {
        case GoalPace.mild:
          rate = 0.10;
          break;
        case GoalPace.standard:
          rate = 0.15;
          break;
        case GoalPace.aggressive:
          rate = 0.20;
          break;
      }
      rawTargetCalories = rawMaintenanceCalories * (1.0 - rate);
    } else if (profile.targetWeightKg > profile.currentWeightKg) {
      final double rate;
      switch (profile.goalPace) {
        case GoalPace.mild:
          rate = 0.05;
          break;
        case GoalPace.standard:
          rate = 0.10;
          break;
        case GoalPace.aggressive:
          rate = 0.15;
          break;
      }
      rawTargetCalories = rawMaintenanceCalories * (1.0 + rate);
    } else {
      rawTargetCalories = rawMaintenanceCalories;
    }

    final rawProteinGrams = profile.currentWeightKg * _proteinGramsPerKg;
    final rawFatGrams = profile.currentWeightKg * _fatGramsPerKg;
    final rawCarbsGrams = (rawTargetCalories -
            rawProteinGrams * _caloriesPerProteinGram -
            rawFatGrams * _caloriesPerFatGram) /
        _caloriesPerCarbGram;

    if (rawBasalCalories <= 0 ||
        rawMaintenanceCalories <= 0 ||
        rawTargetCalories <= 0 ||
        rawCarbsGrams < 0) {
      return const CalculationResult.needsProfessionalReview(
        "Không thể tạo mục tiêu dinh dưỡng an toàn từ dữ liệu hiện tại.",
      );
    }

    return CalculationResult.target(
      NutritionTarget(
        basalCalories: rawBasalCalories.round(),
        maintenanceCalories: rawMaintenanceCalories.round(),
        calories: rawTargetCalories.round(),
        proteinGrams: rawProteinGrams.round(),
        carbsGrams: rawCarbsGrams.round(),
        fatGrams: rawFatGrams.round(),
        audit: NutritionTargetAudit(
          rawBasalCalories: rawBasalCalories,
          rawMaintenanceCalories: rawMaintenanceCalories,
          rawTargetCalories: rawTargetCalories,
          rawProteinGrams: rawProteinGrams,
          rawCarbsGrams: rawCarbsGrams,
          rawFatGrams: rawFatGrams,
        ),
      ),
    );
  }

  int capAutomaticCalorieDelta(int currentCalories, int requestedDelta) {
    if (currentCalories <= 0) return 0;

    final fivePercentCap = (currentCalories * _automaticChangeRate).round();
    final cap = min(fivePercentCap, _maximumAutomaticCalories);
    return requestedDelta.clamp(-cap, cap);
  }

  String? _validate(PersonalProfile profile, int ageYears) {
    if (ageYears < _minimumAge || ageYears > _maximumAge) {
      return "Độ tuổi cần nằm trong khoảng 18 đến 100.";
    }
    if (!profile.heightCm.isFinite ||
        profile.heightCm < _minimumHeightCm ||
        profile.heightCm > _maximumHeightCm) {
      return "Chiều cao nằm ngoài phạm vi hỗ trợ.";
    }
    if (!profile.currentWeightKg.isFinite ||
        profile.currentWeightKg < _minimumWeightKg ||
        profile.currentWeightKg > _maximumWeightKg) {
      return "Cân nặng hiện tại nằm ngoài phạm vi hỗ trợ.";
    }
    if (!profile.targetWeightKg.isFinite ||
        profile.targetWeightKg < _minimumWeightKg ||
        profile.targetWeightKg > _maximumWeightKg) {
      return "Cân nặng mục tiêu nằm ngoài phạm vi hỗ trợ.";
    }
    if (!profile.personalizationConsent) {
      return "Cần đồng ý sử dụng dữ liệu hồ sơ trước khi tính mục tiêu.";
    }
    return null;
  }

  String? _validateTimeline(PersonalProfile profile, TargetTimeline? timeline) {
    if (timeline == null) return null;

    final durationDays = timeline.targetDateEpochDay - timeline.todayEpochDay;
    if (durationDays <= 0) {
      return "Ngày mục tiêu cần nằm sau ngày hiện tại.";
    }

    final durationWeeks = durationDays / _daysPerWeek;
    final kilogramsPerWeek =
        (profile.targetWeightKg - profile.currentWeightKg).abs() / durationWeeks;

    if (kilogramsPerWeek > _maximumKilogramsPerWeek) {
      return "Tốc độ thay đổi cân nặng vượt 0,9 kg mỗi tuần; hãy chọn thời hạn dài hơn hoặc trao đổi với chuyên gia.";
    }
    return null;
  }
}
